test/unit/builtins/promise.spec.ts

```diff
@@ -0,0 +1,1022 @@
+            log("promise2 rejected with: " + reason);
+        });
+
+        resolve("Hello!");
+
+        return allLogs;
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual(["promise2 rejected with: fulfill exception!"]);
+});
+
+test("promise then onRejected throws", () => {
+    util.testFunction`
+        const { promise, reject } = defer<string>();
+
+        const promise2 = promise.then(
+            _ => {},
+            reason => { throw "fulfill exception from onReject!" }
+        );
+
+        promise2.catch(reason => {
+            log("promise2 rejected with: " + reason);
+        });
+
+        reject("Hello!");
+
+        return allLogs;
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual(["promise2 rejected with: fulfill exception from onReject!"]);
+});
+
+test("then on resolved promise immediately calls callback", () => {
+    util.testFunction`
+        Promise.resolve(42).then(data => { log(data); });
+
+        return allLogs;
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual([42]);
+});
+
+test("then on rejected promise immediately calls callback", () => {
+    util.testFunction`
+        Promise.reject("already rejected").then(data => { log("resolved", data); }, reason => { log("rejected", reason); });
+
+        return allLogs;
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual(["rejected", "already rejected"]);
+});
+
+test("catch on rejected promise immediately calls callback", () => {
+    util.testFunction`
+        Promise.reject("already rejected").catch(reason => { log(reason); });
+
+        return allLogs;
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual(["already rejected"]);
+});
+
+test("finally on resolved promise immediately calls callback", () => {
+    util.testFunction`
+        Promise.resolve(42).finally(() => { log("finally"); });
+
+        return allLogs;
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual(["finally"]);
+});
+
+test("finally on rejected promise immediately calls callback", () => {
+    util.testFunction`
+        Promise.reject("already rejected").finally(() => { log("finally"); });
+
+        return allLogs;
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual(["finally"]);
+});
+
+test("example: asynchronous web request", () => {
+    const testHarness = `
+        interface UserData { name: string, age: number}
+        const requests = new Map<number, (userData: UserData) => void>();
+        function getUserData(id: number, callback: (userData: UserData) => void) {
+            requests.set(id, callback);
+        }
+        function emulateRequestReturn(id: number, data: UserData) {
+            requests.get(id)!(data);
+        }
+    `;
+
+    util.testFunction`
+        // Wrap function getUserData(id: number, callback: (userData: UserData) => void) into a Promise
+        async function getUserDataAsync(id: number): Promise<UserData> {
```

Comment:
dumb question: how does this work if we don't support async yet?

Reply:
It works because async keyword has no special function yet and will just be ignored. You are right that this is a a strange async, and in fact I think it is removed again in my async/await branch. I will remove it here too.